buildSrc/src/main/java/net/neoforged/neodev/GenerateAccessTransformers.java

```diff
@@ -0,0 +1,245 @@
+package net.neoforged.neodev;
+
+import org.apache.commons.lang3.mutable.MutableInt;
+import org.gradle.api.DefaultTask;
+import org.gradle.api.file.RegularFileProperty;
+import org.gradle.api.provider.ListProperty;
+import org.gradle.api.tasks.Input;
+import org.gradle.api.tasks.InputFile;
+import org.gradle.api.tasks.OutputFile;
+import org.gradle.api.tasks.TaskAction;
+import org.objectweb.asm.ClassReader;
+import org.objectweb.asm.ClassVisitor;
+import org.objectweb.asm.FieldVisitor;
+import org.objectweb.asm.MethodVisitor;
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.Type;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.io.Serializable;
+import java.nio.file.Files;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.IntStream;
+import java.util.zip.ZipFile;
+
+public abstract class GenerateAccessTransformers extends DefaultTask {
+    public static final Modifier PUBLIC = mod("public", false, Opcodes.ACC_PUBLIC);
+    public static final Modifier PROTECTED = mod("protected", false, Opcodes.ACC_PUBLIC, Opcodes.ACC_PROTECTED);
+
+    @InputFile
+    public abstract RegularFileProperty getInput();
+
+    @OutputFile
+    public abstract RegularFileProperty getAccessTransformer();
+
+    @Input
+    public abstract ListProperty<AtGroup> getGroups();
+
+    @TaskAction
+    public void exec() throws IOException {
+        Map<String, ClassInfo> targets = new HashMap<>();
+        var visitor = visitor(targets);
```

Comment:
Might be a bit cleaner if you move the visitor to a second class in the file, and make the map a field of the visitor.